File: app/src/main/java/com/xiayiye5/recyclerviewadd/java/AddDataAdapter.java
package com.xiayiye5.recyclerviewadd.java;

import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.xiayiye5.recyclerviewadd.R;

import java.util.List;

/**
 * @author DELL
 */
public class AddDataAdapter extends RecyclerView.Adapter<AddDataAdapter.VH> {
    private List<String> addViewed;
    private final int FooterNum = 2;
    private View normalView;

    AddDataAdapter(List<String> addViewed) {
        this.addViewed = addViewed;
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        if (i == FooterNum) {
            View footerView = View.inflate(viewGroup.getContext(), R.layout.activity_footer_view, null);
            return new VH(footerView);
        } else {
            normalView = View.inflate(viewGroup.getContext(), R.layout.activity_normal_view, null);
            return new VH(normalView);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final VH vh, final int i) {
        if (addViewed.size() != i) {
            vh.tvClickAddTextButton.setText(addViewed.get(i));
            vh.tvClickAddTextButton.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    longClickListenerRemove.setLongClickListener(v);
                    return true;
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return addViewed.size() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == addViewed.size()) {
            return FooterNum;
        } else {
            return 1;
        }
    }

    class VH extends RecyclerView.ViewHolder {
        private TextView tvClickAddTextButton;

        VH(@NonNull final View itemView) {
            super(itemView);
            if (itemView == normalView) {
                tvClickAddTextButton = (TextView) itemView.findViewById(R.id.tv_click_add_text);
            } else {
                TextView tvAddTextButton = (TextView) itemView.findViewById(R.id.tv_add_text_button);
                tvAddTextButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(itemView.getContext(), "点击了添加按钮", Toast.LENGTH_SHORT).show();
                        addDataListener.onAddDataListener(getLayoutPosition());
                    }
                });
            }
        }
    }

    private AddDataListener addDataListener;
    private LongClickListenerRemove longClickListenerRemove;

    void setLongClickListenerRemove(LongClickListenerRemove longClickListenerRemove) {
        this.longClickListenerRemove = longClickListenerRemove;
    }

    void setAddDataListener(AddDataListener addDataListener) {
        this.addDataListener = addDataListener;
    }

    interface AddDataListener {
        /**
         * 添加数据的监听方法
         * @param position 添加的位置
         */
        void onAddDataListener(int position);
    }

    interface LongClickListenerRemove {
        /**
         * 长按删除的方法
         * @param view 删除的view
         */
        void setLongClickListener(View view);
    }
}

File: app/src/main/java/com/xiayiye5/recyclerviewadd/kotlin/AddDataAdapter.kt
package com.xiayiye5.recyclerviewadd.kotlin

import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView

import com.xiayiye5.recyclerviewadd.R

/**
 * @author DELL
 */
class AddDataAdapter(private val addViewed: List<String>) :
    RecyclerView.Adapter<AddDataAdapter.VH>() {
    private val normal = 1
    private val footerNum = 2
    private var footerView: View? = null
    private var normalView: View? = null

    internal var addDataListener: AddDataListener? = null
    internal var longClickListenerRemove: LongClickListenerRemove? = null

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): VH {
        if (i == footerNum) {
            footerView = View.inflate(viewGroup.context, R.layout.activity_footer_view, null)
            return VH(footerView!!)
        } else {
            normalView = View.inflate(viewGroup.context, R.layout.activity_normal_view, null)
            return VH(normalView!!)
        }
    }

    override fun onBindViewHolder(vh: VH, i: Int) {
        if (addViewed.size != i) {
            vh.tvClickAddTextButton!!.text = addViewed[i]
            vh.tvClickAddTextButton!!.setOnLongClickListener { v ->
                longClickListenerRemove?.setLongClickListener(v)
                true
            }
        }
    }

    override fun getItemCount(): Int {
        return addViewed.size + 1
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == addViewed.size) {
            footerNum
        } else {
            normal
        }
    }

    inner class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var tvClickAddTextButton: TextView? = null
        var tvAddTextButton: TextView? = null

        init {
            if (itemView === normalView) {
                tvClickAddTextButton =
                    itemView.findViewById<View>(R.id.tv_click_add_text) as TextView
            } else {
                tvAddTextButton = itemView.findViewById<View>(R.id.tv_add_text_button) as TextView
                tvAddTextButton!!.setOnClickListener {
                    Toast.makeText(itemView.context, "点击了添加按钮", Toast.LENGTH_SHORT).show()
                    addDataListener?.onAddDataListener(layoutPosition)
                }
            }
        }
    }

    fun setLongClickListenerRemove(longClickListenerRemove: LongClickListenerRemove) {
        this.longClickListenerRemove = longClickListenerRemove
    }

    fun setAddDataListener(addDataListener: AddDataListener) {
        this.addDataListener = addDataListener
    }

    interface AddDataListener {
        fun onAddDataListener(position: Int)
    }

    interface LongClickListenerRemove {
        fun setLongClickListener(view: View)
    }
}
